### PR TITLE
Making available transparently more TQDM kwargs

### PR DESCRIPTION
**Pull request by LucaCappelletti94**
## Changes
* Introduced a new class called `TQDMSettings` in the `utils` submodule, making it easy to set custom kwargs for the tqdm loading bar.
* Updated documentations and type hinting of the `verbose` kwarg, changed to `Union[int, TQDMSettings]`
* Resolved minor code smells across the code base pointed out by Pylint
* Adresses issue #296 

While the code added is certainly fully executed, the test suite runs in its entirety and therefore I am reasonably convinced the added code has full coverage, I wasn't able to run the provided `make test-coverage` as it raised several rather obscure errors such as `AttributeError: _ARRAY_API not found`. I don't have the time right now to investigate what weird issue may be causing that error.

## Checklist before requesting a review
- [x] Docstrings added/updated in public functions and classes
- [x] Tests added, reasonable test coverage (at least ~90%, `make test-coverage`)
- [x] Sphinx docs added/updated and render properly (`make docs` and see `docs/_build/index.html`)

### CODE REVIEW

**Comment by j-adamczyk** [comment]
One general request - do not use a custom class, rely on dictionary instead. This will change the type of `verbose` to `Union[int, dict]`.

Other required changes:
- change in the `verbose` docstrings
- change in `@validation_params`, probably to `["verbose", dict]`
- add tests (note that `tqdm` writes to stderr)

Most of this can be easily done en masse, e.g. with "Replace in files" in PyCharm.

File: skfp/bases/base_filter.py [hunk 1]
```diff
@@ -6,11 +6,11 @@
 
 import numpy as np
 from joblib import effective_n_jobs
-from rdkit.Chem import Mol
+from rdkit.Chem.rdchem import Mol
```
Comment (j-adamczyk) [inline skfp/bases/base_filter.py:9]: Unnecessary, we use `from rdkit.Chem import Mol` in all files

**Review by j-adamczyk — CHANGES_REQUESTED** [review]

File: skfp/bases/base_filter.py (hunk 1 above)
Comment (LucaCappelletti94) [reply skfp/bases/base_filter.py]: Reverted in commit [f65c196](https://github.com/scikit-fingerprints/scikit-fingerprints/pull/298/commits/f65c196709b448510631846e9511c6f799b97789)

Comment (LucaCappelletti94) [reply skfp/bases/base_filter.py]: The reason for the change was to remove pylint errors ranting about it. I agree it is one of those discutible errors.

**Comment by LucaCappelletti94** [comment]
> One general request - do not use a custom class, rely on dictionary instead. This will change the type of `verbose` to `Union[int, dict]`.
> 
> Other required changes:
> 
> * change in the `verbose` docstrings, e.g. add `"If a dictionary is passed, it is treated as kwargs for `tqdm()`, and can be used to control the progress bar.`
> * change in `@validation_params`, probably to `["verbose", dict]`
> * add tests (note that `tqdm` writes to stderr)
> 
> Most of this can be easily done en masse, e.g. with "Replace in files" in PyCharm.
> 
> `AttributeError: _ARRAY_API not found` looks like a dependency problem with Numpy 2. I checked and it works for me. Did you use `make setup`?

While we can proceed this way, before I do these changes, why do you prefer a dictionary (which may contain all sorts of things) to a class?

**Comment by j-adamczyk** [comment]
> > One general request - do not use a custom class, rely on dictionary instead. This will change the type of `verbose` to `Union[int, dict]`.
> > Other required changes:
> > 
> > * change in the `verbose` docstrings, e.g. add `"If a dictionary is passed, it is treated as kwargs for `tqdm()`, and can be used to control the progress bar.`
> > * change in `@validation_params`, probably to `["verbose", dict]`
> > * add tests (note that `tqdm` writes to stderr)
> > 
> > Most of this can be easily done en masse, e.g. with "Replace in files" in PyCharm.
> > `AttributeError: _ARRAY_API not found` looks like a dependency problem with Numpy 2. I checked and it works for me. Did you use `make setup`?
> 
> While we can proceed this way, before I do these changes, why do you prefer a dictionary (which may contain all sorts of things) to a class?

A valid question, and the reason is maintenance. This class would basically be a wrapper around `tqdm`, and we would have to maintain it. For example, if `tqdm` adds a new argument, we would also have to, or if they deprecate something, we should also do that. In other words, this would result in a very tight coupling between libraries, which I definitely would like to avoid.

Also, validation is not a problem - we won't do any. User should be concerned with what is passed there, since dictionary gives him complete freedom. `tqdm` docs are good, so this is not an issue, and they perform their own validation. It's also a very flexible approach, so if anything in `tqdm` changes, `scikit-fingerprints` will not be affected.

**Comment by LucaCappelletti94** [comment]
Working on this now.

**Comment by LucaCappelletti94** [comment]
I believe I have reasonably addressed your three requests, let me know whether it seems okay.

File: skfp/utils/parallel.py [hunk 2]
```diff
@@ -101,10 +103,22 @@ def run_in_parallel(
...
+        tqdm_settings = {
+            "total": num_batches,
+            "desc": "Processing",
+            "disable": verbose == 0,
+        }
     else:
+        assert isinstance(verbose, dict)
+        tqdm_settings = verbose.copy()
+        tqdm_settings["total"] = num_batches
+        tqdm_settings["disable"] = verbose.get("disable", False)
```
Comment (j-adamczyk) [inline skfp/utils/parallel.py:116]: Note that `assert` can get erased with Python optimizations. It should not be used for production code, just for tests. Use `elif isinstance(verbose, dict)`, and raise ValueError in `else`

File: tests/preprocessing/input_output/data/mol_out.sdf [hunk 3]
```diff
@@ -0,0 +1,3974 @@
+
```
Comment (j-adamczyk) [inline tests/preprocessing/input_output/data/mol_out.sdf:1]: This should probably be in `.gitignore`

File: tests/fingerprints/functional_groups.py [hunk 4]
```diff
@@ -8,7 +8,9 @@
 
 
 def test_functional_groups_bit_fingerprint(smiles_list, mols_list):
-    fg_fp = FunctionalGroupsFingerprint(n_jobs=-1)
+    fg_fp = FunctionalGroupsFingerprint(n_jobs=-1, verbose={
+        "leave": False,
```
Comment (j-adamczyk) [inline tests/fingerprints/functional_groups.py:12]: Isn't this the default option?

File: tests/fingerprints/map.py [hunk 5]
```diff
@@ -87,7 +91,10 @@ def test_map_sparse_count_fingerprint(smallest_smiles_list, smallest_mols_list):
     map_fp = MAPFingerprint(
         variant="count",
         sparse=True,
-        verbose=0,
+        verbose={
+            "dynamic_ncols": False,
+            "disable": True,
+        },
```
Comment (j-adamczyk) [inline tests/fingerprints/map.py:97]: Why?

**Review by j-adamczyk — CHANGES_REQUESTED** [review]

**Comment by j-adamczyk** [comment]
In general, looks very good. I left a few comments and questions.

File: tests/fingerprints/functional_groups.py (hunk 4 above)
Comment (LucaCappelletti94) [reply tests/fingerprints/functional_groups.py:12]: It very much isn't - and the fact that it is not the default option is what made me do the PR in the first place.

File: tests/fingerprints/map.py (hunk 5 above)
Comment (LucaCappelletti94) [reply tests/fingerprints/map.py:97]: To test some different parametrizations within the tests.

Comment (j-adamczyk) [reply tests/fingerprints/map.py:97]: Ok, but for the final version, leave the existing tests unchanged. This need tests covering this specific functionality anyway.

File: skfp/bases/base_substructure_fp.py [hunk 6]
```diff
@@ -43,8 +42,9 @@ class BaseSubstructureFingerprint(BaseFingerprintTransformer):
         Number of inputs processed in each batch. ``None`` divides input data into
         equal-sized parts, as many as ``n_jobs``.
 
-    verbose : int, default=0
+    verbose : int or dict, default=0
         Controls the verbosity when computing fingerprints.
+        If a dictionary is passed, it is treated as kwargs for tqdm(), and can be used to control the progress bar
```
[the remark this answers is not in the document]
Comment (LucaCappelletti94) [reply skfp/bases/base_substructure_fp.py]: Addressed

File: skfp/utils/parallel.py [hunk 7]
```diff
@@ -13,16 +13,17 @@ class ProgressParallel(Parallel):
 
     Parameters
     ----------
-    total : int, default=None
-        Total number of inputs to process in parallel.
+    tqdm_settings: Optional[dict] = None
+        TQDM settings to use for the progress bar.
```
[the remark this answers is not in the document]
Comment (LucaCappelletti94) [reply skfp/utils/parallel.py]: Addressed

File: skfp/utils/parallel.py [hunk 8]
```diff
@@ -71,9 +72,10 @@ def run_in_parallel(
         Whether to flatten the results, e.g. to change list of lists of integers
         into a list of integers.
 
-    verbose : int, default=0
+    verbose : int or dict, default=0
         Controls the verbosity. If higher than zero, progress bar will be shown,
-        tracking the processing of batches.
+        tracking the processing of batches. If `dict` object is provided,
```
[the remark this answers is not in the document]
Comment (LucaCappelletti94) [reply skfp/utils/parallel.py:79]: Addressed

File: skfp/utils/parallel.py [hunk 9]
```diff
@@ -71,9 +72,10 @@ def run_in_parallel(
...
+        it will be used to configure the tqdm progress bar.
```
[the remark this answers is not in the document]
Comment (LucaCappelletti94) [reply skfp/utils/parallel.py]: Addressed

File: tests/preprocessing/input_output/data/mol_out.sdf (hunk 3 above)
Comment (LucaCappelletti94) [reply tests/preprocessing/input_output/data/mol_out.sdf]: Done

File: skfp/utils/parallel.py (hunk 2 above)
Comment (LucaCappelletti94) [reply skfp/utils/parallel.py]: Done

File: tests/fingerprints/functional_groups.py (hunk 4 above)
Comment (LucaCappelletti94) [reply tests/fingerprints/functional_groups.py]: Removed these tests.

File: tests/fingerprints/map.py (hunk 5 above)
Comment (LucaCappelletti94) [reply tests/fingerprints/map.py]: Removed verbose dictionary these tests.

**Comment by LucaCappelletti94** [comment]
Ok, I think I might have addressed all requests.

File: skfp/bases/base_substructure_fp.py [hunk 10]
```diff
@@ -43,8 +42,10 @@ class BaseSubstructureFingerprint(BaseFingerprintTransformer):
         Number of inputs processed in each batch. ``None`` divides input data into
         equal-sized parts, as many as ``n_jobs``.
 
-    verbose : int, default=0
+    verbose : int or ``dict``, default=0
```
Comment (j-adamczyk) [inline skfp/bases/base_substructure_fp.py:45]: You don't need backticks for parameter types, only for actual descriptions. This is one of many weird things in Sphinx and scikit-learn standard for it. Sorry for not being precise about it earlier.

File: tests/fingerprints/functional_groups.py [hunk 11]
```diff
@@ -8,7 +8,7 @@
 
 
 def test_functional_groups_bit_fingerprint(smiles_list, mols_list):
-    fg_fp = FunctionalGroupsFingerprint(n_jobs=-1)
+    fg_fp = FunctionalGroupsFingerprint(n_jobs=-1, verbose=0)
```
Comment (j-adamczyk) [inline tests/fingerprints/functional_groups.py:11]: Leave existing tests as they were. Instead, add more tests in `tests/utils/parallel.py`. You can also compute any fingerprints there, as needed. We try to isolate what we test - those are not truly unit tests, but here they should reasonably isolate testing the given fingerprint.

File: tests/fingerprints/map.py [hunk 12]
```diff
@@ -46,7 +46,9 @@ def test_map_count_fingerprint(smallest_smiles_list, smallest_mols_list):
 def test_map_raw_hashes_fingerprint(smallest_smiles_list, smallest_mols_list):
     map_fp = MAPFingerprint(
         variant="raw_hashes",
-        verbose=0,
+        verbose={
+            "leave": False,
+        },
```
Comment (j-adamczyk) [inline tests/fingerprints/map.py:51]: You can safely remove `verbose` here entirely

File: tests/utils/parallel.py [hunk 13]
```diff
@@ -43,7 +43,7 @@ def test_run_in_parallel_batch_size():
     result_parallel = run_in_parallel(
         func, data, n_jobs=-1, batch_size=1, flatten_results=True
     )
-    assert result_sequential == result_parallel
+    assert result_sequential == result_parallel
```
Comment (j-adamczyk) [inline tests/utils/parallel.py:46]: Why add a space?

File: tests/utils/parallel.py [hunk 14]
```diff
@@ -43,7 +43,7 @@ def test_run_in_parallel_batch_size():
...
```
Comment (j-adamczyk) [inline tests/utils/parallel.py:48]: And here we would need more tests. Probably `leave and maybe `dynamic_ncols` should be tested, by using `capsys`. You can either test it on some simple functions like in `test_progress_parallel` (more like unit tests), or compute fingerprints (more integration test style).

**Review by j-adamczyk — CHANGES_REQUESTED** [review]

**Comment by j-adamczyk** [comment]
A few minor comments left. Only bigger thing is that we should have tests for `leave` or other kwargs for `tqdm` - this should be possible by comparing its outputs, if I understand correctly.

**Comment by LucaCappelletti94** [comment]
I am not sure I understand the nature of these crashes in the Github Actions - do they have anything to do with this PR?

File: skfp/bases/base_substructure_fp.py (hunk 10 above)
Comment (LucaCappelletti94) [reply skfp/bases/base_substructure_fp.py]: Done

File: tests/fingerprints/map.py (hunk 12 above)
Comment (LucaCappelletti94) [reply tests/fingerprints/map.py:51]: Done

File: tests/fingerprints/functional_groups.py (hunk 11 above)
Comment (LucaCappelletti94) [reply tests/fingerprints/functional_groups.py:11]: Restored as it was earlier

**Comment by j-adamczyk** [comment]
> I am not sure I understand the nature of these crashes in the Github Actions - do they have anything to do with this PR?

Yes, notice that pre-commit hooks are failing. If you did not run `make setup`, you may not have pre-commit hooks turned on. They run a bunch of code quality tools (e.g. black, flake8, mypy) before each commit. Looking at mypy errors, you may need to modify a few additional lines that would not work properly with the `dict` type.

**Comment by LucaCappelletti94** [comment]
Poetry hangs indefinitely when I try to run `make setup`. I will look into this when I have time.

**Comment by mjste** [comment]
For some reason I had similar issues with poetry recently. In my case exporting dependencies to requirements.txt and installing them with bare pip worked.

`poetry export --without-hashes --with dev,doc,test > requirements.txt`
`poetry shell`
`pip install -r requirements.txt`
`poetry self add poetry-plugin-export`
`pre-commit install`

**Comment by LucaCappelletti94** [comment]
I will take a stab at it later today

**Comment by j-adamczyk** [comment]
@LucaCappelletti94 I resolved conflicts on your branch and indentation error in grid search docs. Just add a few more tests for outputs with `tqdm` args you're interested in, and this PR will be ready for merge.

**Comment by LucaCappelletti94** [comment]
Ok, thank you for your work. I have been a bit overwhelmed by work on my end so I ended up postponing finishing the pr

**Review by j-adamczyk — APPROVED** [review]

**Comment by j-adamczyk** [comment]
@LucaCappelletti94 I checked and this is basically not testable, or at least I found no way to properly test `tqdm` outputs non-automatically for those keywords. However, having manually testing this, it looks ok. I also added a few smaller fixes, and this PR should be ready now.

File: tests/conftest.py [hunk 15]
```diff
@@ -77,7 +77,6 @@ def _load_test_data_smiles() -> pd.DataFrame:
             try:
                 filepath = os.path.join(str(curr_dir), "hiv_mol.csv.zip")
                 df = pd.read_csv(filepath)
-                break
```
Comment (mjste) [inline tests/conftest.py:80]: Why removing break?

**Review by mjste — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 16]
```diff
@@ -50,6 +50,7 @@ rdkit = "<=2024.3.3"
 scikit-learn = "*"
 scipy = "*"
 tqdm = "*"
+pytest = "^8.3.3"
```
[the remark this answers is not in the document]
Comment (j-adamczyk) [reply pyproject.toml]: My bad, fixed

File: tests/conftest.py (hunk 15 above)
Comment (j-adamczyk) [reply tests/conftest.py]: My mistake, fixed

**Review by mjste — APPROVED** [review]

**Comment by j-adamczyk** [comment]
@LucaCappelletti94 I'm merging the PR, thanks for the contribution.

**Pull request merged**